Source/HTTP/XCurl/XCurlEasyRequest.cpp

```diff
@@ -0,0 +1,342 @@
+#include "pch.h"
+#include "XCurlEasyRequest.h"
+#include "XCurlProvider.h"
+
+namespace xbox
+{
+namespace http_client
+{
+
+XCurlEasyRequest::XCurlEasyRequest(CURL* curlEasyHandle, HCCallHandle hcCall, XAsyncBlock* async)
+    : m_curlEasyHandle{ curlEasyHandle },
+    m_hcCallHandle{ hcCall },
+    m_asyncBlock{ async }
+{
+}
+
+XCurlEasyRequest::~XCurlEasyRequest()
+{
+    curl_easy_cleanup(m_curlEasyHandle);
+    curl_slist_free_all(m_headers);
+}
+
+Result<HC_UNIQUE_PTR<XCurlEasyRequest>> XCurlEasyRequest::Initialize(HCCallHandle hcCall, XAsyncBlock* async)
+{
+    CURL* curlEasyHandle{ curl_easy_init() };
+    if (!curlEasyHandle)
+    {
+        HC_TRACE_ERROR(HTTPCLIENT, "XCurlEasyRequest::Initialize:: curl_easy_init failed");
+        return E_FAIL;
+    }
+
+    http_stl_allocator<XCurlEasyRequest> a{};
+    HC_UNIQUE_PTR<XCurlEasyRequest> easyRequest{ new (a.allocate(1)) XCurlEasyRequest{ curlEasyHandle, hcCall, async } };
+
+    // body (first so we can override things curl "helpfully" sets for us)
+    uint8_t const* body = nullptr;
+    uint32_t bodySize = 0;
+    RETURN_IF_FAILED(HCHttpCallRequestGetRequestBodyBytes(hcCall, &body, &bodySize));
+
+    if (bodySize > 0)
+    {
+        // we set both POSTFIELDSIZE and INFILESIZE because curl uses one or the
+        // other depending on method
+        RETURN_IF_FAILED(easyRequest->SetOpt(CURLOPT_POSTFIELDSIZE, static_cast<long>(bodySize)));
+        RETURN_IF_FAILED(easyRequest->SetOpt(CURLOPT_INFILESIZE, static_cast<long>(bodySize)));
+
+        // read callback
+        RETURN_IF_FAILED(easyRequest->SetOpt(CURLOPT_READFUNCTION, &ReadCallback));
+        RETURN_IF_FAILED(easyRequest->SetOpt(CURLOPT_READDATA, easyRequest.get()));
+    }
+
+    // url & method
+    char const* url = nullptr;
+    char const* method = nullptr;
+    RETURN_IF_FAILED(HCHttpCallRequestGetUrl(hcCall, &method, &url));
+    RETURN_IF_FAILED(easyRequest->SetOpt(CURLOPT_URL, url));
+
+    CURLoption opt = CURLOPT_HTTPGET;
+    RETURN_IF_FAILED(MethodStringToOpt(method, opt));
```

Comment:
are you sure you don't need to support `DELETE`?

Reply:
we should support it.  Will add.